# Analysis 분석 순서
## 1. Train_data (9973 경기 = 99730 row) *Encoding = utf8
1) 승패에 영향을 주는 지표 (Kill, Death, Assist, Damage 등) 탐색하기

2) EDA를 거치면서 분석에 영향을 줄만한 지표를 만들기 (Ward, KDA2)

3) 앞에서 구한 지표에 머신러닝을 이용하여 Weight와 Bias 학습하기 {model : argmax[ X(1x5) \* W(5x2) + b(1x2) ]}

4) 기타
- 최종 발표 당시) 발표시점에서는 약 6800경기를 이용하여 Kill, Death, Assist, Damage, Ward, KDA2 지표를 학습함.
- 주의) Encoding=utf8으로 되어있어서 엑셀을 사용할 경우 파일이 깨져서 보일 수 있음. (메모장 이용을 추천)
- 주의) Encoding=utf8이 아닌 파일을 read_csv로 열 경우 오류가 생김

### 2. Test_data (1308 경기 = 13080 row) *Encoding = 기본값
1) 머신러닝 결과를 이용해서 테스트를 진행

2) 기타

- 3) Predict와 차이점은 Test는 이미 결과가 나온 경기를 이용하나 3) Predict는 어떤 경기를 기준으로 게임이 시작하기 전이라고 보고 평균 지표를 계산하여 예측한 후 그 경기 결과와 비교하는 것임.
- 주의) Encoding=기본값으로 되어있어서 read_csv에서 utf8로 Encoding된 것도 열 수 있으나 깨진 상태로 열림

### 3. Predict (Our Goal) - (아직 일어나지 않는 경기를 만들어서 예측)
1) 초기 목표는 프로게이머들의 개인전적, 대회전적 등을 분석해서 각 선수가 게임의 승패에 미치는 영향을 알아내어 이적시장에서 최적의 팀을 구성하는 것이 목표였으나, 현재는 여러가지 문제로 진행이 되지 않고, 최상위권 아마추어 유저들과 프로유저들의 평균 지표를 계산하여 이를 앞에서 머신러닝을 통해 학습한 Weight와 Bias를 가지고 예측해보는 것을 진행함.

2) 설명
- 최종 발표 당시) 26개의 무작위로 선정한 경기를 기준으로 정해진 값이 아닌 모든 지표를 게임에 참여한 선수들의 최근 20경기 내의 전적의 평균 지표를 이용하여 계산하였고, 이를 가지고 평가함.
- 최종 발표 당시) Test1(K,D,A,W) : 정답 14, 오답 7, 무승부 5
- 최종 발표 당시) Test2(W,Da,KDA2) : 정답 14, 오답 6, 무승부 6
- 정답은 결과를 정확하게 예측한 경우, 오답은 결과를 정확하게 틀린 경우, 데이터를 양쪽 같은 포지션 선수들의 지표의 차이를 가지고 학습을 하는데, 거기에 -를 모두 곱한 경우(경기가 승패를 기준으로 대칭구조이므로 뒤집은 경우) 결과를 넣어서 한 경기를 2번 예측시켜서 나온 결과를 사용
- EX) (승) 선수1, 선수2, 선수3, 선수4, 선수5 - (패) 선수6, 선수7, 선수8, 선수9, 선수10
- EX) (패) 선수6, 선수7, 선수8, 선수9, 선수10 - (승) 선수1, 선수2, 선수3, 선수4, 선수5
- EX) 학습에 사용된 지표 차: 선수1 and 선수6 (same position in game), (지표1 = 선수1.지표1 - 선수6.지표1) and (지표1' = 선수6.지표1 - 선수1.지표1) 이렇게 1경기마다 뒤집어서 2번 넣음

In [1]:
import tensorflow as tf
import numpy as np
import pandas as pd
import glob

## 1. Train data (9973 경기 = 99730 row)
### 0) Data Load (Train_data.zip 3,043KB)

In [169]:
# get data file names
path =r'C:\Users\waf\Documents\00_DGIST 수업\03_02\Big data\프로젝트\2017_BigData\Data\all_data_utf'
filenames = glob.glob(path + "/*.csv")
filenames
dfs = []
for filename in filenames:
    dfs.append(pd.read_csv(filename, encoding="utf8", engine='python'))
# Concatenate all data into one DataFrame
kda_data = pd.concat(dfs, ignore_index=True)
kda_data['Death_fake'] = kda_data['Death']
kda_data['Death_fake'] = kda_data['Death_fake'].replace(0, 1)
kda_data['KDA'] = (kda_data['Kill'] + kda_data['Assist'])/ kda_data['Death_fake']
kda_data = kda_data.drop('Death_fake', 'columns')
kda_data

Unnamed: 0 MatchNumber GameResult   Champion         Nickname Kill  \
0              0       30220         패배       마오카이         twice손채영    7   
1              1       30220         패배       세주아니         작전동 노란머리    6   
2              2       30220         패배        야스오         hf ghkyu   10   
3              3       30220         패배          진         BtC RYAN    3   
4              4       30220         패배      미스 포츈           커소리어스나    3   
5              5       30220         승리         카밀            Heiz4    5   
6              6       30220         승리        니달리           담원 크러쉬   11   
7              7       30220         승리  트위스티드 페이트       TeachersWu   13   
8              8       30220         승리       이즈리얼          해강고원딜 팬   10   
9              9       30220         승리        카르마            Heart    4   
10            10       30221         승리         카밀            Heiz4    3   
11            11       30221         승리         렝가         twice손채영    7   
12            12       30221         승리       다이애나              이야공    8   
13            13       30221         승리          진         hf ghkyu    7   
14            14       30221         승리       탐 켄치          YserAoA    2   
15            15       30221         패배        제이스          우에하라 크롱    0   
16            16       30221         패배       세주아니              F0y    2   
17            17       30221         패배       오리아나       TeachersWu    1   
18            18       30221         패배       드레이븐          해강고원딜 팬    6   
19            19       30221         패배          쉔        BtC Joker    0   
20            20       30222         승리        제이스           삶이힘든사람    3   
21            21       30222         승리        니달리      sanspareil1    4   
22            22       30222         승리        라이즈         Yaharong    6   
23            23       30222         승리          진  LPL noob ad QAQ   12   
24            24       30222         승리        쓰레쉬         hf ghkyu    5   
25            25       30222         패배      문도 박사             선비모드    5   
26            26       30222         패배        렉사이             마이엠브    3   
27            27       30222         패배         조이         PawN God    3   
28            28       30222         패배      트리스타나       7nian xing    4   
29            29       30222         패배        레오나          pa13tte    1   
...          ...         ...        ...        ...              ...  ...   
99700        170      152146         패배         나르            노니노니노    1   
99701        171      152146         패배        카직스           죽지마얘들아    9   
99702        172      152146         패배        카사딘              스눈이    5   
99703        173      152146         패배      미스 포츈             Grue    1   
99704        174      152146         패배       이즈리얼          후으으으에에엥    6   
99705        175      152146         승리         오른      Gazuaaaaaaa   12   
99706        176      152146         승리        리 신        우지나 나 갸영이    4   
99707        177      152146         승리        야스오              두 런   10   
99708        178      152146         승리          진          구름꽃길만걷자    6   
99709        179      152146         승리        카르마  iwasBornForThis    4   
99710        180      152147         패배         카밀            노니노니노    3   
99711        181      152147         패배        리 신           쌀국수뚝붸기    5   
99712        182      152147         패배        루시안          구름꽃길만걷자    4   
99713        183      152147         패배       칼리스타          후으으으에에엥    4   
99714        184      152147         패배        쓰레쉬         FrostOvO    1   
99715        185      152147         승리        제이스    CantlostDream    5   
99716        186      152147         승리     자르반 4세              두 런   11   
99717        187      152147         승리        르블랑         KDM Edge   12   
99718        188      152147         승리      미스 포츈     pecko phase2    6   
99719        189      152147         승리         잔나        우지나 나 갸영이    1   
99720        190      152148         승리         나르    

### 1) 승패에 영향을 줄만한 지표 탐색
- 각 팀별로 특정 지표의 단순 합을 계산하여 높은 쪽이 이긴다고 예측하여 지표를 탐색
- more_win : 지표의 합이 높은 쪽이 이긴 경우
- more_lose : 지표의 합이 높은 쪽이 진 경우
- not_more : 지표의 합이 같은 경우

In [24]:
def Sum_predict(parameter):
    more_win = 0
    more_lose = 0
    not_more = 0
    total = 0
    tmp = []
    index = 0
    for i in kda_data.groupby(['MatchNumber','GameResult'])[parameter].sum():
        tmp.append(i)
        index += 1
        if index == 2:
            total += 1
            if tmp[0] > tmp[1]:
                more_win += 1
            elif tmp[0] < tmp[1]:
                more_lose += 1
            else:
                not_more += 1
            tmp = []
            index = 0
    # print(more_kill_win/total)
    print("more win  :",more_win/total)
    print("more lose :",more_lose/total)
    print("not more  :",not_more/total)

### 1-1) Kill Sum

In [25]:
Sum_predict("Kill")

more win  : 0.9465557003910559
more lose : 0.039506668003609746
not more  : 0.013937631605334403


### 1-2) Assist Sum

In [27]:
Sum_predict("Assist")

more win  : 0.9294094053945653
more lose : 0.06357164343728065
not more  : 0.007018951168154016


### 1-3) KDA Sum [(K+A)/D]

In [28]:
Sum_predict("KDA")

more win  : 0.9495638223202647
more lose : 0.05043617767973529
not more  : 0.0


### 1-4) Damage Sum

In [30]:
Sum_predict("Damage")

more win  : 0.8752632106688057
more lose : 0.12473678933119423
not more  : 0.0


### 1-5) Buy_Pward Sum

In [31]:
Sum_predict("Buy_Pward")

more win  : 0.5449714228416725
more lose : 0.3916574751829941
not more  : 0.0633711019753334


### 1-6) Build_ward

In [33]:
Sum_predict("Build_ward")

more win  : 0.571041812894816
more lose : 0.4058959189812494
not more  : 0.023062268123934623


### 1-7) Destroy_ward

In [34]:
Sum_predict("Destroy_ward")

more win  : 0.5921989371302517
more lose : 0.3620776095457736
not more  : 0.04572345332397473


### 1-8) CS

In [35]:
Sum_predict("CS")

more win  : 0.814699689160734
more lose : 0.17978542063571643
not more  : 0.005514890203549584


### 1-9) Baron

In [36]:
Sum_predict("Baron")

more win  : 0.5298305424646546
more lose : 0.06838463852401484
not more  : 0.4017848190113306


### 1-10) Dragon

In [37]:
Sum_predict("Dragon")

more win  : 0.6211771783816304
more lose : 0.158828837862228
not more  : 0.21999398375614157


### 1-11) Tower

In [38]:
Sum_predict("Tower")

more win  : 0.9666098465857816
more lose : 0.01574250476285972
not more  : 0.017647648651358667


### 2) 분석에 도움이 될만한 새로운 지표 만들기

### 2-1) KDA2 Sum [(K+0.2A-1.2D] (몇 번의 시도 끝에 가장 높다고 생각하는 비율을 만듬)

In [171]:
kda_data['KDA2'] = kda_data['Kill'] + 0.2*kda_data['Assist'] - 1.2*kda_data['Death']
Sum_predict("KDA2")

more win  : 0.9582873759149704
more lose : 0.04151208262308232
not more  : 0.0002005414619472576


### 2-2) Ward (0.18\*Buy_Pward+Build_ward+2\*Destroy_ward) [몇번의 시도 끝에 높은 수치]

In [172]:
kda_data["Ward"] = 0.18*kda_data["Buy_Pward"] + kda_data["Build_ward"] + 2*kda_data["Destroy_ward"]
Sum_predict("Ward")

more win  : 0.7212473678933119
more lose : 0.2782512784518199
not more  : 0.000501353654868144


### 2-3) Object (3.2\*Baron+Dragon) [몇번의 시도 끝에 높은 수치]

In [173]:
kda_data["Object"] = 3.2*kda_data["Baron"] + 1*kda_data["Dragon"]
Sum_predict("Object")

more win  : 0.763361074902236
more lose : 0.11932216985861827
not more  : 0.1173167552391457


### 3) 앞에서 탐색한 지표로 머신러닝으로 W, b 구하기
### 3-1) data preprocess

In [174]:
sub = kda_data[["MatchNumber","GameResult","Kill","Death","Assist","Ward","Damage","KDA2","Tower","Object","CS"]]
result = []

for i in sub["GameResult"]:
    if i == "승리":
        result.append(1)
    else:
        result.append(0)
        
sub["Result"] = result;
sub.head(20)

C:\Users\waf\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.


MatchNumber GameResult Kill Death Assist   Ward Damage KDA2 Tower Object  \
0        30220         패배    7     9      6  28.36  20924 -2.6     2      1   
1        30220         패배    6     7     11  41.08  12273 -0.2     2      1   
2        30220         패배   10     7      7  16.54  23123    3     2      1   
3        30220         패배    3    11      6  22.18   9015   -9     2      1   
4        30220         패배    3     9     12  42.26  16067 -5.4     2      1   
5        30220         승리    5     9      9     29  20328   -4    10    4.2   
6        30220         승리   11     4     13   34.9  23256  8.8    10    4.2   
7        30220         승리   13     3      8  23.72  24899   11    10    4.2   
8        30220         승리   10     7     15  21.18  32430  4.6    10    4.2   
9        30220         승리    4     6     19   31.9  15573  0.6    10    4.2   
10       30221         승리    3     2      9     11  12685  2.4     8    5.2   
11       30221         승리    7     2      9  48.54  10853  6.4     8    5.2   
12       30221         승리    8     0      7   7.54  15203  9.4     8    5.2   
13       30221         승리    7     4      7     14  11205  3.6     8    5.2   
14       30221         승리    2     1     14  31.08   4112  3.6     8    5.2   
15       30221         패배    0     6      3  35.44   8506 -6.6     3      0   
16       30221         패배    2     4      5  37.72   2902 -1.8     3      0   
17       30221         패배    1     5      3  19.54   5786 -4.4     3      0   
18       30221         패배    6     6      0  13.36  11433 -1.2     3      0   
19       30221         패배    0     6      6  41.44   5647   -6     3      0   

     CS  Result  
0   149       0  
1   130       0  
2   210       0  
3   164       0  
4    42       0  
5   196       1  
6   141       1  
7   255       1  
8   146       1  
9    58       1  
10  126       1  
11  193       1  
12  187       1  
13  143       1  
14   34       1  
15  122       0  
16  114       0  
17  180       0  
18  165       0  
19   39       0

In [134]:
def sigmoid(n):
    return 1 / (1 + np.exp(-n))

In [82]:
def cvt(parameter):
    index = 0
    result = []
    tmp = []
    for item in sub[parameter]:
        index += 1
        tmp.append(item)
        if index % 10 == 0:
            result.append(tmp)
            tmp=[]
    result = np.array(result)
    res = np.ones([result.shape[0],int(result.shape[1]/2)])
    for i in range(len(result)):
        res[i] = result[i][0:5]-result[i][5:10]
    return res

In [106]:
def cvt_sum(parameter): # 10명의 전체 합에 대한 비율 * 100
    index = 0
    result = []
    tmp = []
    for item in sub[parameter]:
        index += 1
        tmp.append(item)
        if index % 10 == 0:
            sum_da = 0
            for j in tmp:
                sum_da += j
            result.append(np.array(tmp)/sum_da*100)
            tmp=[]
    result = np.array(result)
    res = np.ones([result.shape[0],int(result.shape[1]/2)])
    for i in range(len(result)):
        res[i] = result[i][0:5]-result[i][5:10]
    return res

In [ ]:
def cvt_result(parameter): # 결과를 뽑을 때 사용
    index = 0
    result = []
    tmp = []
    for item in sub[parameter]:
        index += 1
        if index % 10 == 1:
            tmp.append(item)
        if index % 10 == 6:
            tmp.append(item)
        if index % 10 == 0:
            result.append(tmp)
            tmp=[]
    result = np.array(result)
    return result

In [175]:
kill = cvt("Kill")
death = cvt("Death")
assist = cvt("Assist")
kda2 = cvt("KDA2")
ward = cvt_sum("Ward")
damage = cvt_sum("Damage")
tower = cvt("Tower")
obj = cvt("Object")
CS = cvt_sum("CS")
ylabel = cvt_result("Result")
actual = np.argmax(ylabel,axis=1)
actual.shape

(9973,)

### 3-2) Machine Learning
- X는 10명의 데이터를 5명, 5명 팀으로 나눈 후 같은 포지션의 지표를 뺀 값
- argmax(X(1x5) \* W(5x2) + b(1x2))

In [126]:
trait = 3000 # 학습 횟수
train_list = [kill,death,assist,kda2,ward,damage,tower,obj,CS]
for i,data in enumerate(train_list):
    x_data = data
    y_data = ylabel
    
    X = tf.placeholder(tf.float32)
    Y = tf.placeholder(tf.float32)

    W = tf.Variable(tf.random_uniform([5, 2], -1., 1.))
    b = tf.Variable(tf.zeros([2]))

    L = tf.add(tf.matmul(X, W), b)
    L = tf.nn.relu(L)
    model = tf.nn.softmax(L)

    cost = tf.reduce_mean(-tf.reduce_sum(Y * tf.log(model), axis=1))
    optimizer = tf.train.AdamOptimizer(learning_rate=0.01)
    train_op = optimizer.minimize(cost)

    init = tf.global_variables_initializer()
    sess = tf.Session()
    sess.run(init)
    for step in range(trait):
        sess.run(train_op, feed_dict={X: x_data, Y: y_data})
        
        if (step + 1) == 1:
            print(step + 1, "cost:", sess.run(cost, feed_dict={X: x_data, Y: y_data}))
        if (step + 1) % (trait/10) == 0:
            print(step + 1, "cost:", sess.run(cost, feed_dict={X: x_data, Y: y_data}))

    prediction = tf.argmax(model, 1)
    target = tf.argmax(Y, 1)
    print('예측값:', sess.run(prediction, feed_dict={X: x_data}))
    print('실제값:', sess.run(target, feed_dict={Y: y_data}))
    
    is_correct = tf.equal(prediction, target)
    accuracy = tf.reduce_mean(tf.cast(is_correct, tf.float32))
    print('정확도: %.2f' % sess.run(accuracy * 100, feed_dict={X: x_data, Y: y_data}))
    
    if i == 0:
        w1 = sess.run(W)
        b1 = sess.run(b)
#         a1 = sess.run(accuracy)
    elif i == 1:
        w2 = sess.run(W)
        b2 = sess.run(b)
#         a2 = sess.run(accuracy)
    elif i == 2:
        w3 = sess.run(W)
        b3 = sess.run(b)
#         a3 = sess.run(accuracy)
    elif i == 3:
        w4 = sess.run(W)
        b4 = sess.run(b)
#         a4 = sess.run(accuracy)
    elif i == 4:
        w5 = sess.run(W)
        b5 = sess.run(b)
#         a5 = sess.run(accuracy)
    elif i == 5:
        w6 = sess.run(W)
        b6 = sess.run(b)
#         a6 = sess.run(accuracy)
    elif i == 6:
        w7 = sess.run(W)
        b7 = sess.run(b)
#         a7 = sess.run(accuracy)
    elif i == 7:
        w8 = sess.run(W)
        b8 = sess.run(b)
#         a8 = sess.run(accuracy)
    elif i == 8:
        w9 = sess.run(W)
        b9 = sess.run(b)
#         a9 = sess.run(accuracy)

1 cost: 1.46767
300 cost: 0.172476
600 cost: 0.134929
900 cost: 0.133602
1200 cost: 0.133003
1500 cost: 0.132717
1800 cost: 0.132485
2100 cost: 0.132443
2400 cost: 0.132443
2700 cost: 0.132443
3000 cost: 0.132443
예측값: [1 0 0 ..., 1 1 0]
실제값: [1 0 0 ..., 1 1 0]
정확도: 95.44
1 cost: 0.45803
300 cost: 0.14046
600 cost: 0.139554
900 cost: 0.139521
1200 cost: 0.139521
1500 cost: 0.139521
1800 cost: 0.139521
2100 cost: 0.139521
2400 cost: 0.139521
2700 cost: 0.139521
3000 cost: 0.139521
예측값: [1 0 0 ..., 1 1 0]
실제값: [1 0 0 ..., 1 1 0]
정확도: 95.14
1 cost: 0.878693
300 cost: 0.194885
600 cost: 0.191024
900 cost: 0.190831
1200 cost: 0.190831
1500 cost: 0.190831
1800 cost: 0.190831
2100 cost: 0.190831
2400 cost: 0.190831
2700 cost: 0.190831
3000 cost: 0.190831
예측값: [1 0 0 ..., 1 1 0]
실제값: [1 0 0 ..., 1 1 0]
정확도: 93.31
1 cost: 1.7505
300 cost: 0.115569
600 cost: 0.114591
900 cost: 0.114312
1200 cost: 0.114261
1500 cost: 0.114251
1800 cost: 0.114251
2100 cost: 0.114251
2400 cost: 0.114251
2700 cost: 0

In [177]:
predict = []
for i in range(int(len(sub)/10)):
    predict.append(np.argmax(0.9544*sigmoid(np.matmul(kill[i],w1)+b1)+0.9514*sigmoid(np.matmul(death[i],w2)+b2)
                             +0.9331*sigmoid(np.matmul(assist[i],w3)+b3)+0.9608*sigmoid(np.matmul(kda2[i],w4)+b4)
                             +0.7308*sigmoid(np.matmul(ward[i],w5)+b5)+0.8963*sigmoid(np.matmul(damage[i],w6)+b6)
                             +0.9740*sigmoid(np.matmul(tower[i],w7)+b7)+0.8267*sigmoid(np.matmul(obj[i],w8)+b8)
                             +0.8258*sigmoid(np.matmul(CS[i],w9)+b9)))

score = 0
for i in range(len(predict)):
    if predict[i] == actual[i]:
        score += 1
score = score/len(predict)
print("Train_data 정확도 :",score)

Train_data 정확도 : 0.9719241953273839


## 2. Test_data (1308 경기 = 13080 row)
- 머신러닝 결과 확인용
- Train_data와 Encoding이 다름을 주의

In [162]:
# get data file names
path =r'C:\Users\waf\Documents\00_DGIST 수업\03_02\Big data\프로젝트\2017_BigData\Data\all_data'
filenames = glob.glob(path + "/*.csv")
filenames
dfs = []
for filename in filenames:
    dfs.append(pd.read_csv(filename, engine='python'))
# Concatenate all data into one DataFrame
test_data = pd.concat(dfs, ignore_index=True)
test_data['Death_fake'] = test_data['Death']
test_data['Death_fake'] = test_data['Death_fake'].replace(0, 1)
test_data['KDA'] = (test_data['Kill'] + test_data['Assist'])/ test_data['Death_fake']
test_data = test_data.drop('Death_fake', 'columns')
test_data

Unnamed: 0  MatchNumber GameResult Champion          Nickname  Kill  \
0               0       154169         패배       오른               Fly     0   
1               1       154169         패배      리 신        1112213311     2   
2               2       154169         패배      아지르             웨이 지안     3   
3               3       154169         패배    미스 포츈             ARE2A     5   
4               4       154169         패배        쉔          혼자가 편해졌어     2   
5               5       154169         승리     갱플랭크         bbq Crazy     5   
6               6       154169         승리      이블린     jinwandalaohu    12   
7               7       154169         승리       조이      whisper snow     0   
8               8       154169         승리      바루스       Labixiaoxin     7   
9               9       154169         승리      트런들          라인전이기니갱와     1   
10             10       154170         패배     이렐리아            고개를 젓다     1   
11             11       154170         패배   자르반 4세       Kiendeptrai     1   
12             12       154170         패배      신드라        1112213311     0   
13             13       154170         패배     이즈리얼           영광 스런 년     0   
14             14       154170         패배       타릭          라인전이기니갱와     0   
15             15       154170         승리       잭스           my head     6   
16             16       154170         승리      쉬바나  so nihao zaijian     4   
17             17       154170         승리      라이즈           나에게 기회를     3   
18             18       154170         승리      시비르          처음은누구나초보     0   
19             19       154170         승리      카르마       Sim Student     0   
20             20       154171         승리     신 짜오     Ningxia Allah    11   
21             21       154171         승리      카직스          mcmglllc     5   
22             22       154171         승리      제라스        1112213311     8   
23             23       154171         승리     케이틀린             ARE2A     3   
24             24       154171         승리     마오카이       Sim Student     4   
25             25       154171         패배      야스오          Pz ZZang     5   
26             26       154171         패배     세주아니      Diamondprox1     5   
27             27       154171         패배      베이가        TeachersWu     6   
28             28       154171         패배    트리스타나          처음은누구나초보     6   
29             29       154171         패배      레오나    Gambit Support     2   
...           ...          ...        ...      ...               ...   ...   
13050         170       153531         패배      제이스              혼자혼자     0   
13051         171       153531         패배      쉬바나          WENJUNXI     2   
13052         172       153531         패배      말자하               에어맨     2   
13053         173       153531         패배        진            991126     2   
13054         174       153531         패배        쉔           스칼렛 십자군     3   
13055         175       153531         승리      사이온          IGS Bear     2   
13056         176       153531         승리     세주아니            수지Suzy     6   
13057         177       153531         승리     갱플랭크             이재훈다딱     7   
13058         178       153531         승리    미스 포츈   Gambit Marksman     2   
13059         179       153531         승리     모르가나    Gambit Support     2   
13060         180       153532         패배       나르              혼자혼자     2   
13061         181       153532         패배   자르반 4세       yuanyouyuan     2   
13062         182       153532         패배      아지르          게임좋아공부싫어     4   
13063         183       153532         패배     이즈리얼       Labixiaoxin     1   
13064         184       153532         패배        쉔         ollehKing     2   
13065         185       153532         승리       카밀             pppyp     5   
13066         186       153532         승리      카직스       Sorrymybad2     7   
13067         187       153532         승리       조이               수 노    14   
13068         188       153532         승리        진           형섭 사랑해요     7   
13069         189       15353

In [163]:
test_data["Object"] = 3.2*test_data["Baron"] + 1*test_data["Dragon"]
test_data["Ward"] = 0.18*test_data["Buy_Pward"] + test_data["Build_ward"] + 2*test_data["Destroy_ward"]
test_data['KDA2'] = test_data['Kill'] + 0.2*test_data['Assist'] - 1.2*test_data['Death']
sub_test = test_data[["MatchNumber","GameResult","Kill","Death","Assist","Ward","Damage","KDA2","Tower","Object","CS"]]
result = []

for i in sub_test["GameResult"]:
    if i == "승리":
        result.append(1)
    else:
        result.append(0)
        
sub_test["Result"] = result;
sub_test.head(20)

C:\Users\waf\Anaconda3\lib\site-packages\ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  del sys.path[0]


MatchNumber GameResult  Kill  Death  Assist   Ward  Damage  KDA2  Tower  \
0        154169         패배     0      6       3   9.18   10743  -6.6      1   
1        154169         패배     2      3       5  39.44    6993  -0.6      1   
2        154169         패배     3      4       1  20.54   10188  -1.6      1   
3        154169         패배     5      8       4  18.18   15458  -3.8      1   
4        154169         패배     2      4       9  21.90    5339  -1.0      1   
5        154169         승리     5      0      10  28.36   21246   7.0      5   
6        154169         승리    12      2       4  20.18   12514  10.4      5   
7        154169         승리     0      1       9  28.54   10661   0.6      5   
8        154169         승리     7      3       7  18.72    9957   4.8      5   
9        154169         승리     1      6      13  44.08    5358  -3.6      5   
10       154170         패배     1      4       0   6.18    4417  -3.8      0   
11       154170         패배     1      4       0  13.18    6462  -3.8      0   
12       154170         패배     0      2       0   9.18    3891  -2.4      0   
13       154170         패배     0      2       0   9.00    5555  -2.4      0   
14       154170         패배     0      1       0   7.18     636  -1.2      0   
15       154170         승리     6      0       1   9.00    8684   6.2      8   
16       154170         승리     4      1       3  11.90    5277   3.4      8   
17       154170         승리     3      1       2  14.18    6816   2.2      8   
18       154170         승리     0      0       4  11.36    4732   0.8      8   
19       154170         승리     0      0       7  15.90    3276   1.4      8   

    Object   CS  Result  
0      0.0  112       0  
1      0.0  101       0  
2      0.0  187       0  
3      0.0  150       0  
4      0.0   45       0  
5      0.0  221       1  
6      0.0   91       1  
7      0.0  156       1  
8      0.0  183       1  
9      0.0   47       1  
10     0.0  124       0  
11     0.0   68       0  
12     0.0  117       0  
13     0.0  121       0  
14     0.0   24       0  
15     1.0  152       1  
16     1.0  119       1  
17     1.0  159       1  
18     1.0  145       1  
19     1.0    7       1

In [164]:
def cvt2(parameter):
    index = 0
    result = []
    tmp = []
    for item in sub_test[parameter]:
        index += 1
        tmp.append(item)
        if index % 10 == 0:
            result.append(tmp)
            tmp=[]
    result = np.array(result)
    res = np.ones([result.shape[0],int(result.shape[1]/2)])
    for i in range(len(result)):
        res[i] = result[i][0:5]-result[i][5:10]
    return res

In [165]:
def cvt_sum2(parameter): # 10명의 전체 합에 대한 비율 * 100
    index = 0
    result = []
    tmp = []
    for item in sub_test[parameter]:
        index += 1
        tmp.append(item)
        if index % 10 == 0:
            sum_da = 0
            for j in tmp:
                sum_da += j
            result.append(np.array(tmp)/sum_da*100)
            tmp=[]
    result = np.array(result)
    res = np.ones([result.shape[0],int(result.shape[1]/2)])
    for i in range(len(result)):
        res[i] = result[i][0:5]-result[i][5:10]
    return res

In [166]:
def cvt_result2(parameter): # 결과를 뽑을 때 사용
    index = 0
    result = []
    tmp = []
    for item in sub_test[parameter]:
        index += 1
        if index % 10 == 1:
            tmp.append(item)
        if index % 10 == 6:
            tmp.append(item)
        if index % 10 == 0:
            result.append(tmp)
            tmp=[]
    result = np.array(result)
    return result

In [167]:
kill = cvt2("Kill")
death = cvt2("Death")
assist = cvt2("Assist")
kda2 = cvt2("KDA2")
ward = cvt_sum2("Ward")
damage = cvt_sum2("Damage")
tower = cvt2("Tower")
obj = cvt2("Object")
CS = cvt_sum2("CS")
ylabel = cvt_result2("Result")
actual = np.argmax(ylabel,axis=1)
actual.shape

(1308,)

- 앞의 계수는 머신러닝 결과 정확도를 넣어준 것, 변수로 받아서 자동화 하려고 했으나 문제가 발생했고, 시간상 손으로 붙임

In [168]:
predict = []
for i in range(int(len(sub_test)/10)):
    predict.append(np.argmax(0.9544*sigmoid(np.matmul(kill[i],w1)+b1)+0.9514*sigmoid(np.matmul(death[i],w2)+b2)
                             +0.9331*sigmoid(np.matmul(assist[i],w3)+b3)+0.9608*sigmoid(np.matmul(kda2[i],w4)+b4)
                             +0.7308*sigmoid(np.matmul(ward[i],w5)+b5)+0.8963*sigmoid(np.matmul(damage[i],w6)+b6)
                             +0.9740*sigmoid(np.matmul(tower[i],w7)+b7)+0.8267*sigmoid(np.matmul(obj[i],w8)+b8)
                             +0.8258*sigmoid(np.matmul(CS[i],w9)+b9)))

score = 0
for i in range(len(predict)):
    if predict[i] == actual[i]:
        score += 1
score = score/len(predict)
print("Test_data 정확도 :",score)

Test_data 정확도 : 0.9571865443425076


In [ ]:
# get data file names
filename = "match35.csv"
test_data = pd.read_csv(filename, engine='python')
test_data['Death_fake'] = test_data['Death']
test_data['Death_fake'] = test_data['Death_fake'].replace(0, 1)
test_data['KDA'] = (test_data['Kill'] + test_data['Assist'])/ test_data['Death_fake']
test_data = test_data.drop('Death_fake', 'columns')
test_data["Team"] = [1,1,1,1,1,2,2,2,2,2]
test_data

In [ ]:
predict = []
for i in range(int(len(sub)/10)):
    predict.append(np.argmax(sigmoid(np.matmul(kill[i], w1) + b1)+sigmoid(np.matmul(death[i], w2) + b2)+sigmoid(np.matmul(assist[i], w3) + b3)))

In [ ]:
score = 0
for i in range(len(predict)):
    if predict[i] == actual[i]:
        score += 1
score = score/len(predict)

## 3. Predict
### 1) 개인전적를 임의로 뽑아서 평균 지표를 넣고 머신러닝 결과를 적용 (26 경기)
- 머신러닝 정확도
- Train_data : 0.9719241953273839
- Test_data : 0.9571865443425076
- Predict : 0.6153846153846154
- Predict는 몇 개의 요소를 빼가면서 어떤 것이 중요하나 탐색하였고 그 결과 KDA2와 Ward 점수가 예측에 큰 영향을 주는 것을 확인

In [178]:
# get data file names
path =r'C:\Users\waf\Documents\00_DGIST 수업\03_02\Big data\프로젝트\2017_BigData\Data\match\match\일반'
filenames = glob.glob(path + "/*.csv")
filenames
dfs = []
for filename in filenames:
    dfs.append(pd.read_csv(filename, engine='python'))
# Concatenate all data into one DataFrame
test_data = pd.concat(dfs, ignore_index=True)
test_data['Death_fake'] = test_data['Death']
test_data['Death_fake'] = test_data['Death_fake'].replace(0, 1)
test_data['KDA'] = (test_data['Kill'] + test_data['Assist'])/ test_data['Death_fake']
test_data = test_data.drop('Death_fake', 'columns')
test_data["Team"] = [1,1,1,1,1,2,2,2,2,2]*len(dfs)
test_data

Assist     Baron  Build_ward  Buy_Pward          CS  CSperMin  \
0     4.518519  0.425926   19.185185   2.166667  189.129630  7.227778   
1     6.388889  0.305556   24.000000   1.666667  122.555556  4.702778   
2     5.270270  0.270270   15.081081   2.918919  169.675676  6.754054   
3     7.625000  0.531250   22.937500   4.906250  198.250000  7.759375   
4     7.968750  0.687500   10.562500   1.375000  183.250000  5.887500   
5     7.395349  0.372093   13.860465   1.720930  191.232558  7.539535   
6    10.548387  0.258065   19.129032   4.548387  115.483871  4.864516   
7     8.900000  0.233333   22.233333   4.366667  217.766667  8.630000   
8     5.463415  0.390244   13.439024   1.756098  194.780488  7.751220   
9     8.051282  0.410256   16.205128   4.794872  172.461538  6.017949   
10    7.200000  0.700000   19.733333   1.733333  221.766667       NaN   
11    7.371429  0.714286   25.257143   4.542857  160.942857       NaN   
12    6.903226  0.741935   17.967742   2.967742  206.387097       NaN   
13    8.700000  1.000000   22.300000   1.400000  213.200000       NaN   
14   10.250000  0.708333   25.541667   3.604167   58.187500       NaN   
15    4.477273  0.272727   11.863636   1.159091  168.068182       NaN   
16    7.966667  0.333333   33.166667   5.100000  138.433333       NaN   
17    6.260870  0.282609   22.913043   2.934783  207.869565       NaN   
18    6.152174  0.260870   16.239130   3.108696  196.673913       NaN   
19    9.326087  0.347826   24.304348   6.804348   44.521739       NaN   
20    6.500000  0.843750   19.750000   2.031250  221.812500       NaN   
21    8.285714  0.464286   24.785714   6.142857  137.428571       NaN   
22    7.096774  0.612903   25.870968   3.806452  255.193548       NaN   
23   12.666667  1.000000   31.833333   2.416667  287.500000       NaN   
24    8.392857  0.642857   19.214286   4.285714  137.964286       NaN   
25    6.297297  0.486486   25.837838   3.297297  251.297297       NaN   
26    4.766667  0.400000   16.433333   3.100000  184.366667       NaN   
27    6.500000  0.470588   22.970588   5.558824  256.117647       NaN   
28    8.833333  0.433333   32.200000   7.266667  241.100000       NaN   
29    5.678571  0.178571   34.107143   5.571429  127.142857       NaN   
..         ...       ...         ...        ...         ...       ...   
230   8.304348  0.336957   12.793478   2.315217  120.445652       NaN   
231   6.230769  0.516484   19.637363   3.054945  160.901099       NaN   
232   7.288136  0.525424   14.152542   3.813559  197.915254       NaN   
233   6.390000  0.550000    9.350000   1.410000  167.360000       NaN   
234  10.946565  0.519084   30.610687   8.717557   62.122137       NaN   
235   4.612903  0.387097   11.096774   0.661290  206.129032       NaN   
236   9.715686  0.450980   14.215686   5.313725  140.156863       NaN   
237   5.914286  0.314286   16.528571   3.800000  158.414286       NaN   
238   5.602564  0.307692   12.679487   2.974359  158.871795       NaN   
239  10.125000  0.427083   41.197917   5.302083   59.270833       NaN   
240   5.090909  0.469697   13.409091   1.530303  176.090909       NaN   
241   8.586207  0.327586   23.827586   5.120690  148.275862       NaN   
242   5.315068  0.328767   19.342466   3.726027  192.219178       NaN   
243   5.846154  0.507692   20.323077   4.215385  238.384615       NaN   
244   5.976744  0.244186   13.593023   3.488372  117.500000       NaN   
245   5.555556  0.648148   20.555556   4.435185  207.453704       NaN   
246   8.313253  0.578313   28.783133   5.722892  129.927711       NaN   
247   6.622222  0.600000   17.644444   4.533333  223.388889       NaN   
248   9.037037  0.456790   25.432099   4.370370  139.753086       NaN   
249   7.862069  0.724138   15.873563   4.091954  191.402299       NaN   
250   4.863014  0.246575    8.164384   0.739726  184.547945       NaN   
251   6.871795  0.346154   15.448718   0.807692  163.179487       NaN   
252   5.914286  0.314286   16.528571   3.800000  158.414286       

In [179]:
test_data["Object"] = 3.2*test_data["Baron"] + 1*test_data["Dragon"]
test_data["Ward"] = 0.18*test_data["Buy_Pward"] + test_data["Build_ward"] + 2*test_data["Destroy_ward"]
test_data['KDA2'] = test_data['Kill'] + 0.2*test_data['Assist'] - 1.2*test_data['Death']
sub_test = test_data[["Team","Kill","Death","Assist","Ward","Damage","KDA2","Tower","Object","CS"]]
sub_test.head(20)

Team      Kill     Death     Assist       Ward        Damage      KDA2  \
0      1  4.259259  4.833333   4.518519  33.834444  13707.833333 -0.637037   
1      1  5.416667  6.138889   6.388889  41.077778  11653.888889 -0.672222   
2      1  2.972973  4.567568   5.270270  31.336216  13428.378378 -1.454054   
3      1  5.156250  5.156250   7.625000  38.633125  20548.218750  0.493750   
4      1  6.000000  6.500000   7.968750  28.185000  18383.000000 -0.206250   
5      2  4.720930  3.558140   7.395349  21.193488  14046.837209  1.930233   
6      2  5.580645  4.806452  10.548387  34.528387  11955.161290  1.922581   
7      2  4.100000  2.566667   8.900000  38.686000  18663.566667  2.800000   
8      2  6.804878  3.268293   5.463415  24.291707  17078.731707  3.975610   
9      2  4.717949  5.000000   8.051282  28.863077  16949.076923  0.328205   
10     1  5.933333  3.066667   7.200000  31.778667  18523.266667  3.693333   
11     1  5.000000  4.971429   7.371429  46.646286  13011.142857  0.508571   
12     1  4.741935  3.870968   6.903226  34.630968  18445.548387  1.477419   
13     1  5.000000  1.300000   8.700000  34.952000  19894.600000  5.180000   
14     1  3.916667  4.791667  10.250000  41.440417  11350.687500  0.216667   
15     2  3.227273  5.500000   4.477273  17.981364  12414.772727 -2.477273   
16     2  4.766667  3.966667   7.966667  55.751333  11358.100000  1.600000   
17     2  4.326087  3.282609   6.260870  38.093478  19105.173913  1.639130   
18     2  3.217391  3.217391   6.152174  27.494348  12758.521739  0.586957   
19     2  1.021739  4.478261   9.326087  41.746522   6018.391304 -2.486957   

       Tower    Object          CS  
0   3.962963  2.048148  189.129630  
1   3.861111  1.366667  122.555556  
2   4.000000  1.675676  169.675676  
3   4.250000  2.731250  198.250000  
4   4.812500  3.262500  183.250000  
5   6.348837  2.423256  191.232558  
6   5.612903  1.567742  115.483871  
7   5.933333  1.780000  217.766667  
8   7.243902  2.639024  194.780488  
9   6.128205  2.235897  172.461538  
10  7.233333  3.773333  221.766667  
11  5.571429  3.571429  160.942857  
12  6.967742  4.083871  206.387097  
13  8.700000  5.000000  213.200000  
14  6.145833  3.475000   58.187500  
15  4.068182  1.736364  168.068182  
16  4.000000  1.800000  138.433333  
17  3.239130  1.513043  207.869565  
18  3.652174  1.530435  196.673913  
19  3.543478  2.026087   44.521739

- ylabel은 파일로 받는 걸로 했으나 시간상으로 문제로 일단은 손으로 기록

In [180]:
kill = cvt2("Kill")
death = cvt2("Death")
assist = cvt2("Assist")
kda2 = cvt2("KDA2")
ward = cvt_sum2("Ward")
damage = cvt_sum2("Damage")
tower = cvt2("Tower")
obj = cvt2("Object")
CS = cvt_sum2("CS")
actual = [1, 0, 0, 0, 0, 0,
          1, 0, 1, 1, 0,
          0, 0, 1, 1, 0,
          0, 0, 1, 0, 1,
          1, 0, 1, 1, 1]

In [ ]:
    predict.append(np.argmax(0.9544*sigmoid(np.matmul(kill[i],w1)+b1)+0.9514*sigmoid(np.matmul(death[i],w2)+b2)
                             +0.9331*sigmoid(np.matmul(assist[i],w3)+b3)+0.9608*sigmoid(np.matmul(kda2[i],w4)+b4)
                             +0.7308*sigmoid(np.matmul(ward[i],w5)+b5)+0.8963*sigmoid(np.matmul(damage[i],w6)+b6)
                             +0.9740*sigmoid(np.matmul(tower[i],w7)+b7)+0.8267*sigmoid(np.matmul(obj[i],w8)+b8)
                             +0.8258*sigmoid(np.matmul(CS[i],w9)+b9)))

In [199]:
predict = []
for i in range(int(len(sub_test)/10)):
    predict.append(np.argmax(0.9608*sigmoid(np.matmul(kda2[i],w4)+b4)+0.7308*sigmoid(np.matmul(ward[i],w5)+b5)))

score = 0
for i in range(len(predict)):
    if predict[i] == actual[i]:
        score += 1
score = score/len(predict)
print("Predict 정확도 :",score)

Predict 정확도 : 0.6153846153846154


### 2) 응용 분야 : 프로게이머 전적 데이터를 이용하여 팀끼리 경기를 시켜서 나오는 결과 예측 (4개팀, 6경기)